ops/main.py

```diff
@@ -356,9 +325,255 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
         return False
 
 
+@dataclasses.dataclass(frozen=True)
+class _CharmSpec:
+    """Charm spec."""
+
+    charm_type: Type["ops.charm.CharmBase"]
+    charm_root: Path
+    meta: Optional[Dict[str, Any]]
+    actions: Optional[Dict[str, Any]] = None
+    config: Optional[Dict[str, Any]] = None
+
+    @staticmethod
+    def _get_charm_dir():
+        charm_dir = os.environ.get("JUJU_CHARM_DIR")
+        if charm_dir is None:
+            # Assume $JUJU_CHARM_DIR/lib/op/main.py structure.
+            charm_dir = Path(f'{__file__}/../../..').resolve()
+        else:
+            charm_dir = Path(charm_dir).resolve()
+        return charm_dir
+
+    @staticmethod
+    def autoload(charm_type: Type["ops.charm.CharmBase"]):
+        charm_root = _CharmSpec._get_charm_dir()
+
+        metadata_path = charm_root / "metadata.yaml"
+        if not metadata_path.exists():
+            raise FileNotFoundError(
+                f"invalid charm root {charm_root!r}; "
+                f"expected to contain at least a `metadata.yaml` file.",
+            )
+        meta = yaml.safe_load(metadata_path.open())
+
+        actions = config = None
+
+        config_path = charm_root / "config.yaml"
+        if config_path.exists():
+            config = yaml.safe_load(config_path.open())
+
+        actions_path = charm_root / "actions.yaml"
+        if actions_path.exists():
+            actions = yaml.safe_load(actions_path.open())
+
+        return _CharmSpec(
+            charm_type=charm_type,
+            charm_root=charm_root,
+            meta=meta,
+            actions=actions,
+            config=config
+        )
+
+
+class Ops:
```

Comment:
@benhoyt and I feel that it would be best to start this out as `_Ops`, rather than being a public interface immediately. It wouldn't be hard to move from `_Ops` to `Ops` in the future, whereas once it's public it's very difficult to go back. It also allows a bit more freedom adjusting in the public members of the class at first, while still providing the nicer abstraction and easier overriding that refactoring into the class provides.